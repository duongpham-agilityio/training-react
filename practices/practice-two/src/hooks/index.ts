export * from './useBookForm';
export * from './useSearchParam';
export * from './usePagination';
export * from './useFilter';
export * from './useDebounce';
export * from './useBooks';
export * from './useBook';
export * from './usePopupContext';
export * from './useToastContext';
