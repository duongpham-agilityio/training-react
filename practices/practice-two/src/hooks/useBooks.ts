import { useCallback } from 'react';
import useSWR from 'swr';
import {
  useFilter,
  usePagination,
  usePopupContext,
  useSearchParam,
  useToastContext,
} from 'hooks';

// Helpers
import { axiosConfig } from 'helpers';

// Types
import { Book } from 'types';

// Constants
import { ENDPOINT, MESSAGES } from '@constants';

/**
 * Get out all the current list
 * @returns object containing properties and methods for interacting with a books
 */
export const useBooks = () => {
  const { dispatch } = usePopupContext();
  const { setNotification } = useToastContext();
  const swr = useSWR<Book[]>(`${ENDPOINT.BOOKS}${ENDPOINT.SORT}`);
  const {
    param: { name, sort, page },
    ...restUseSearch
  } = useSearchParam();
  const { data: filters } = useFilter(swr.data || [], {
    name,
    sort,
  });
  const pagination = usePagination(filters, page);
  const handleDelete = useCallback(
    (id: number) => {
      try {
        dispatch(() => {
          const newData = swr.data?.filter((book) => {
            return book.id !== id;
          });

          swr.mutate(() => axiosConfig.delete(`${ENDPOINT.BOOKS}/${id}`), {
            optimisticData: newData,
            populateCache: false,
            revalidate: false,
          });

          setNotification({ message: 'Delete success', title: 'Delete ' });
        });
      } catch (error) {
        return setNotification({
          message: MESSAGES.ERROR_TITLE,
          title: MESSAGES.EMPTY_FIELD,
          type: 'error',
        });
      }
    },
    [dispatch, setNotification, swr]
  );

  return {
    ...swr,
    ...restUseSearch,
    ...pagination,
    param: { name, sort, page },
    deleteBook: handleDelete,
  };
};
