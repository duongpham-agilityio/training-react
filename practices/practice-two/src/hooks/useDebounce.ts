import { useMemo, useRef } from 'react';

// Constants
import { TIMEOUT_DEBOUNCE } from '@constants';

/**
 * - This is a custom hook that listens for user input
 * @param callback Have a string as parameter
 * @returns
 */
export const useDebounce = (callback?: (_value: string) => void) => {
  const refTime = useRef<ReturnType<typeof setTimeout>>();

  const debounce = useMemo(() => {
    return (value: string) => {
      if (refTime.current) clearTimeout(refTime.current);

      refTime.current = setTimeout(() => {
        if (callback) callback(value.trim());
      }, TIMEOUT_DEBOUNCE);
    };
  }, [callback]);

  return debounce;
};
