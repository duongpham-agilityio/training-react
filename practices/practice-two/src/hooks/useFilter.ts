import { SORT } from '@constants';
import { useMemo } from 'react';
import { Book } from 'types';

export type Filter = {
  name: string;
  sort: string;
};

/**
 * Filter books on demand
 * @param data book list
 * @param filter filter option
 * @returns book list after filtering
 */
export const useFilter = (data: Book[], filter: Filter) => {
  /**
   * Sort books on demand
   */
  const sorts = useMemo(() => {
    const newBooks = [...data];
    if (!filter.sort) return data;

    if (filter.sort === SORT.ASCENDING) {
      newBooks.sort(function (a, b) {
        if (a.name < b.name) return -1;

        if (a.name > b.name) return 1;

        return 0;
      });

      return newBooks;
    }

    newBooks.sort(function (a, b) {
      if (a.name < b.name) return 1;

      if (a.name > b.name) return -1;

      return 0;
    });

    return newBooks;
  }, [data, filter.sort]);

  /**
   * Filter by book title or author name
   */
  const filters = useMemo(() => {
    const books = sorts.filter((book) => {
      const isName = book.name
        .toLowerCase()
        .includes(filter.name.toLowerCase());
      const isAuthor = book.author
        .toLowerCase()
        .includes(filter.name.toLowerCase());

      return isName || isAuthor;
    });

    return books;
  }, [sorts, filter.name]);

  return { data: filters };
};
