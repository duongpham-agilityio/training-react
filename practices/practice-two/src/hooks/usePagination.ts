import { useMemo } from 'react';

// Types
import { Book } from 'types';

// Constants
import { RECORD } from '@constants';

export const usePagination = (products: Book[] = [], page: number) => {
  /**
   * Calculate the number of pages to RECORD
   */
  const pagination = useMemo(() => {
    const filtersLength = products.length;
    const isSizePage = filtersLength % RECORD;
    const sizePage = Math.floor(filtersLength / RECORD);

    if (!isSizePage) {
      return new Array(sizePage).fill(0);
    }

    return new Array(sizePage + 1).fill(0);
  }, [products]);

  /**
   * Get products by page
   */
  const filters = useMemo(() => {
    const filteredProducts = products.filter((product) => {
      const index = products.indexOf(product);
      const isStartIndex = index >= (page - 1) * RECORD;
      const isEndIndex = index < page * RECORD;
      const condition = isStartIndex && isEndIndex;

      return condition;
    });

    return filteredProducts;
  }, [products, page]);

  return {
    data: filters,
    pagination,
  };
};
