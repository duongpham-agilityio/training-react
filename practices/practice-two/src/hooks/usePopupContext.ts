import { useContext } from 'react';

// Constants
import { MESSAGES } from '@constants';

// Contexts
import { PopupContext } from 'contexts/Popup/context';

export const usePopupContext = () => {
  const context = useContext(PopupContext);

  if (!context) throw MESSAGES.CONTEXT_ERROR;

  return context;
};
