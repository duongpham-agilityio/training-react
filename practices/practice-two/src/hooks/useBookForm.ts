/* eslint-disable no-irregular-whitespace */
/* eslint-disable @typescript-eslint/no-explicit-any */
import {
  ChangeEvent,
  FormEvent,
  useCallback,
  useMemo,
  useRef,
  useState,
} from 'react';

// Hook
import { useToastContext } from 'hooks';

// Helpers
import {
  axiosConfig,
  convertStringToTime,
  convertTimeToDate,
  validate,
} from 'helpers';

// Services
import { uploadImage } from 'services';

// Constants
import { MESSAGES, TIMEOUT_DEBOUNCE } from '@constants';

// Mock data
import { book } from 'mock-data';

// Types
import { Book, Recommend } from 'types';

export type FormData = Omit<
  Book,
  'publishDate' | 'deletedAt' | 'createdAt' | 'updatedAt'
> & {
  publishDate?: number;
};

/**
 * Interacting with forms
 * @param data initialization values â€‹of data fields in form
 * @param type To know if this is a create form or an update form
 * @returns Object
 */
export const useBookForm = (
  data: FormData,
  type: 'create' | 'update',
  submitHandler: (data: FormData) => Promise<void>
) => {
  const { setNotification } = useToastContext();
  const [state, setState] = useState<FormData & { imageName: string }>({
    ...data,
    imageName: '',
  });
  const [isUpload, setIsUpload] = useState(false);
  const [isSubmit, setIsSubmit] = useState(false);
  const [booksRecommended, setBooksRecommended] = useState<Recommend[]>([]);
  const refImage = useRef<HTMLInputElement>(null);
  const refTime = useRef<ReturnType<typeof setTimeout>>();
  /**
   * Convert data to render form
   */
  const value = useMemo(() => {
    const { publishDate, ...rest } = state;

    return {
      ...rest,
      publishDate: publishDate ? convertTimeToDate(publishDate) : '',
    };
  }, [state]);

  /**
   * Handling get recommend
   */
  const handleRecommended = useCallback((value: string) => {
    if (value) {
      if (refTime.current) clearTimeout(refTime.current);

      refTime.current = setTimeout(async () => {
        try {
          const books = await axiosConfig
            .get(`${process.env.VITE_RECOMMENDED_URL}?q=${value}`)
            .then((r) => r.data)
            .then((data) => {
              // eslint-disable-next-line @typescript-eslint/no-unused-vars
              const books =
                data?.items.map((_item: { volumeInfo: any }) => {
                  const { title, publishedDate, description, imageLinks } =
                    _item.volumeInfo;

                  return {
                    name: title,
                    publishDate: publishedDate,
                    description,
                    imageURL: imageLinks.thumbnail || book.imageURL,
                  };
                }) || [];

              return books;
            });

          setBooksRecommended(books);
        } catch (error) {
          setBooksRecommended([]);
        }
      }, TIMEOUT_DEBOUNCE);
    }
  }, []);

  /**
   * Handling select an recommend
   */
  const handleSelectRecommended = useCallback(
    (index: number) => {
      const book = booksRecommended[index];

      setState((prev) => ({
        ...prev,
        ...book,
      }));
      setBooksRecommended([]);
    },
    [booksRecommended]
  );

  const onSubmit = useCallback(
    (event: FormEvent) => {
      event.preventDefault();
      const { name, author, imageURL, description, publishDate } = state;
      const checked = {
        name,
        author,
        imageURL,
        description,
        publishDate,
      };
      const isError = validate(checked);

      if (isError) {
        return setNotification({
          message: MESSAGES.EMPTY_FIELD_DESCRIPTION,
          title: MESSAGES.EMPTY_FIELD,
          type: 'error',
        });
      }

      setIsSubmit(true);
      submitHandler(state).finally(() => setIsSubmit(false));
    },
    [setNotification, state, submitHandler]
  );

  const onChange = useCallback(
    async (event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
      const element = event.target;
      const value = element.value;
      const key: keyof typeof data = element.name as keyof typeof data;

      if (key === 'publishDate') {
        return setState((prev) => ({
          ...prev,
          publishDate: convertStringToTime(value.trim()),
        }));
      }

      if (key === 'imageURL') {
        const imageEl = element as HTMLInputElement;
        const files = imageEl.files;

        if (files) {
          const data = new FormData();
          const file = files[0];

          data.append('image', file);
          setIsUpload(true);
          uploadImage(data, (response) => {
            if (response.status) {
              const name = file.name;

              setIsUpload(false);
              setState((prev) => ({
                ...prev,
                imageURL: response.data.url,
                imageName: name,
              }));

              return;
            }

            return setNotification({
              message: MESSAGES.ERROR_TITLE,
              title: MESSAGES.EMPTY_FIELD,
              type: 'error',
            });
          });

          return;
        }

        return;
      }

      if (key === 'name') handleRecommended(value.trim());

      setState((prev) => ({
        ...prev,
        [key]: value,
      }));
    },
    [handleRecommended, setNotification]
  );

  return {
    value,
    isUpload,
    booksRecommended,
    refImage,
    isSubmit,
    handleSelectRecommended,
    resetRecommended: setBooksRecommended,
    onSubmit,
    onChange,
  };
};
