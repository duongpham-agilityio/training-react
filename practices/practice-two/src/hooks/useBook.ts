import { useCallback } from 'react';
import { useParams } from 'react-router-dom';
import useSWR from 'swr';

// Types
import { Book } from 'types';

// Helpers
import { axiosConfig } from 'helpers';

// Constants
import { ENDPOINT } from '@constants';
import { usePopupContext } from './usePopupContext';

/**
 * Get the data of a book
 * @returns object containing properties and methods for interacting with a book
 */
export const useBook = () => {
  const { dispatch } = usePopupContext();
  const { id } = useParams();
  const { mutate, ...rest } = useSWR<Book>(`${ENDPOINT.BOOKS}/${id}`);

  const deleteBook = useCallback(() => {
    dispatch(() => {
      mutate(
        async () => {
          return await axiosConfig.delete(`${ENDPOINT.BOOKS}/${id}`);
        },
        {
          populateCache: false,
          revalidate: false,
        }
      );

      window.location.replace(`${ENDPOINT.BOOKS}`);
    });
  }, [dispatch, id, mutate]);

  return {
    ...{ ...rest, mutate },
    deleteBook,
  };
};
