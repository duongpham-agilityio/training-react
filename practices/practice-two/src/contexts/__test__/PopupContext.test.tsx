import { fireEvent, render } from '@testing-library/react';

// Hooks
import { usePopupContext } from 'hooks';

// Mocks
import 'components/__test__/mocks/image.test';

// Providers
import PopupProvider from 'contexts/Popup';

const Child = () => {
  const { dispatch } = usePopupContext();

  return <button onClick={() => dispatch(() => {})}>Open popup</button>;
};
const setup = () =>
  render(<Child />, {
    wrapper: PopupProvider,
  });

describe('Popup context', () => {
  it('Match to snapshot', () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it('Render with props', () => {
    const { getByText, getByRole } = setup();
    const openBtn = getByRole('button');

    fireEvent.click(openBtn);

    const acceptBtn = getByText('Ok');

    fireEvent.click(acceptBtn);

    expect(acceptBtn).not.toBeInTheDocument();
  });

  it('Close popup', () => {
    const { getByText, getByRole } = setup();
    const openBtn = getByRole('button');

    fireEvent.click(openBtn);

    const cancelBtn = getByText('Cancel');

    fireEvent.click(cancelBtn);

    expect(cancelBtn).not.toBeInTheDocument();
  });
});
