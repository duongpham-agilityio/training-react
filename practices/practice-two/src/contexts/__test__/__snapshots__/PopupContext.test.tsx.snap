// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup context Match to snapshot 1`] = `
<div>
  <button>
    Open popup
  </button>
</div>
`;
